index.d.ts

```diff
@@ -0,0 +1,208 @@
+/// <reference types="node" />
+/// <reference types="aws-lambda" />
+/// <reference types="connect" />
+
+import { IncomingMessage, ServerResponse } from 'http';
+import { Handler } from 'aws-lambda';
+import { ErrorHandleFunction } from 'connect';
+
+export default agent;
+
+declare const agent: Agent;
+
+declare class Agent implements Taggable, StartSpanFn {
+  // Configuration
+  start (options?: AgentConfigOptions): Agent;
+  isStarted (): boolean;
+
+  // Data collection hooks
+  middleware: { connect (): ErrorHandleFunction };
+  lambda (handler: Handler): Handler;
+  lambda (type: string, handler: Handler): Handler;
+  handleUncaughtExceptions (
+    fn: (err: Error) => void
+  ): void;
+
+  // Errors
+  captureError (
+    err: Error | string | ParameterizedMessageObject,
+    callback?: CaptureErrorCallback
+  ): void;
+  captureError (
+    err: Error | string | ParameterizedMessageObject,
+    options?: CaptureErrorOptions,
+    callback?: CaptureErrorCallback
+  ): void;
+
+  // Transactions
+  startTransaction (
+    name?: string,
+    type?: string,
+    options?: TransactionOptions
+  ): Transaction | null;
+  setTransactionName (name: string): void;
+  endTransaction (result?: string, endTime?: number): void; // TODO: Should we allow number as well for result?
+  currentTransaction: Transaction | null;
+
+  // Spans
+  startSpan (name?: string, type?: string, options?: SpanOptions): Span | null;
+  currentSpan: Span | null;
+
+  // Context
+  setTag (name: string, value: TagValue): boolean;
+  addTags (tags: Tags): boolean;
+  setUserContext (user: UserObject): void;
+  setCustomContext (custom: object): void;
+
+  // Transport
+  addFilter (fn: FilterFn): void;
+  addErrorFilter (fn: FilterFn): void;
+  addSpanFilter (fn: FilterFn): void;
+  addTransactionFilter (fn: FilterFn): void;
+  flush (callback: Function): void;
+  destroy (): void;
+
+  // Utils
+  logger: Logger;
+}
+
+declare class GenericSpan implements Taggable {
+  // TODO: The following should not be documented right? constructor(), timestamp, ended, id, traceId, parentId, sampled, duration()
```

Comment:
We don't currently have docs for timestamp, ended, id, traceId, parentId and sampled, so I think omitting them is fine. We could have a _separate_ discussion if it's a good idea to _make_ some of those public, but that could just be a follow-up PR.